sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_client.hpp

```diff
@@ -114,53 +114,96 @@ namespace Azure { namespace Security { namespace Attestation {
    */
 
   class AttestationClient final {
+    /** @brief Construct a new Attestation Client object
+     *
+     * @param endpoint The URL address where the client will send the requests to.
+     * @param credential The authentication method to use (required for TPM attestation).
+     * @param options The options to customize the client behavior.
+     */
+    AttestationClient(
+        std::string const& endpoint,
+        std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+        AttestationClientOptions options = AttestationClientOptions());
+
   public:
     /**
      * @brief Destructor.
      *
      */
     virtual ~AttestationClient() = default;
 
+    /** @brief Construct a new Attestation Client object from an existing attestation client.
+     *
+     * @param attestationClient An existing attestation client.
+     */
+    AttestationClient(AttestationClient const& attestationClient)
+        : m_endpoint(attestationClient.m_endpoint), m_apiVersion(attestationClient.m_apiVersion),
+          m_pipeline(attestationClient.m_pipeline),
+          m_tokenValidationOptions(attestationClient.m_tokenValidationOptions){};
+
     /** @brief Construct a new Attestation Client object
      *
      * @param endpoint The URL address where the client will send the requests to.
      * @param credential The authentication method to use (required for TPM attestation).
      * @param options The options to customize the client behavior.
+     * @return AttestationClient The newly created client.
      */
-    explicit AttestationClient(
+    static AttestationClient const Create(
         std::string const& endpoint,
         std::shared_ptr<Core::Credentials::TokenCredential const> credential,
-        AttestationClientOptions options = AttestationClientOptions());
+        AttestationClientOptions options = AttestationClientOptions(),
+        Azure::Core::Context const& constext = Azure::Core::Context{});
 
     /** @brief Construct a new anonymous Attestation Client object
      *
      * @param endpoint The URL address where the client will send the requests to.
      * @param options The options to customize the client behavior.
+     * @return AttestationClient The newly created client.
      *
      * @note TPM attestation requires an authenticated attestation client.
      */
-    explicit AttestationClient(
+    static AttestationClient const Create(
```

Comment:
I have some feelings here about returning a `const` object.  Why did you decide this?

To me, as a user, it means I am getting an instance that I can't mutate or call any non-constant method from it.  Is that expected?

I would typically expect to be the customer who would decide if calling:
`auto client = AttestationClient::Create();`
or
`auto const client = AttestationClient::Create();`

depending on my intentions for the client.  But it might be an explanation

Reply:
The C++ SDK guidelines say that clients are immutable.

The way you express immutability in C++ is to mark the class as `const`. I'm leveraging language features to make the contract specified in the guidelines explicit.

The downside of that is const poisoning. But there is no ambiguity on the part of customers that service clients are immutable.

Reply:
Actually I was wrong. Most of the other languages guidelines say clients are immutable, C++ is mute on the subject. Given @JeffreyRichter's clarification that the real requirement is thread safe and not immutable, I'm reverting this change.